File: docs/demo/V4.x/TESS_PythonAPI_demo/V4公众号案例代码/多时段信控/MyNet.py
import json
from typing import Dict, List
from DLLs.Tessng import PyCustomerNet, tessngIFace, Online, m2p, p2m


class MyNet(PyCustomerNet):
    def __init__(self):
        super(MyNet, self).__init__()
        self.iface = tessngIFace()
        self.netiface = self.iface.netInterface()
        self.simuiface = self.iface.simuInterface()

        self.signal_controller_id_mapping: Dict[int, int] = {}
        self.signal_phase_id_mapping: Dict[int, int] = {}

    def afterLoadNet(self):
        light_data = json.load(open("light_data.json", encoding="utf-8"))

        # 创建信号机、信控方案、信号相位
        for signal_controller_data in light_data["signalControllers"]:
            signal_controller_id: int = signal_controller_data["id"]
            signal_controller_name: str = signal_controller_data["name"]
            # 创建信号机
            signal_controller = self.netiface.createSignalController(signal_controller_name)
            self.signal_controller_id_mapping[signal_controller_id] = signal_controller.id()
            print(f"创建了信号机，id={signal_controller.id()}")
            for signal_plan_data in signal_controller_data["signalPlans"]:
                signal_plan_id: int = signal_plan_data["id"]
                signal_plan_name: str = signal_plan_data["name"]
                cycle_time: int = signal_plan_data["cycleTime"]
                start_time: int = signal_plan_data["startTime"]
                end_time: int = signal_plan_data["endTime"]
                offset: int = signal_plan_data["offset"]
                # 创建信控方案
                signal_plan = self.netiface.createSignalPlan(
                    signal_controller, signal_plan_name, cycle_time, offset, start_time, end_time
                )
                print(f"创建了信控方案，id={signal_plan.id()}")
                for signal_phase_data in signal_plan_data["signalPhases"]:
                    signal_phase_id: int = signal_phase_data["id"]
                    signal_phase_name: str = signal_phase_data["name"]
                    colors: List[str] = signal_phase_data["colors"]
                    durations: List[int] = signal_phase_data["durations"]
                    # 创建信号相位
                    color_intervals = [
                        Online.ColorInterval(color, duration)
                        for color, duration in zip(colors, durations)
                    ]
                    signal_phase = self.netiface.createSignalPlanSignalPhase(
                        signal_plan, signal_phase_name, color_intervals
                    )
                    self.signal_phase_id_mapping[signal_phase_id] = signal_phase.id()
                    print(f"创建了信号相位，id={signal_phase.id()}")
        print()

        # 创建信号灯头
        for signal_lamp in light_data["signalLamps"]:
            signal_lamp_name: str = signal_lamp["name"]
            link_ids: List[int] = signal_lamp["linkIds"]
            signal_controller_id: int = signal_lamp["signalControllerId"]
            signal_controller_tess_id: int = self.signal_controller_id_mapping[signal_controller_id]
            signal_controller = self.netiface.findSignalControllerById(signal_controller_tess_id)
            signal_phase_ids: List[int] = signal_lamp["signalPhaseIds"]
            signal_phase_tess_ids: List[int] = [
                self.signal_phase_id_mapping[signal_phase_id]
                for signal_phase_id in signal_phase_ids
            ]

            for link_id in link_ids:
                link = self.netiface.findLink(link_id)
                for lane in link.lanes():
                    lane_id = lane.id()
                    lane_length = p2m(lane.length())
                    dist = m2p(lane_length - 1)
                    signal_lamp = self.netiface.createTrafficSignalLamp(
                        signal_controller, signal_lamp_name, lane_id, -1, dist
                    )
                    print(f"创建了信号灯头，id={signal_lamp.id()}")
                    for signal_phase_id in signal_phase_tess_ids:
                        self.netiface.addSignalPhaseToLamp(signal_phase_id, signal_lamp)

File: docs/demo/V4.x/TESS_JavaAPI_demo/Traffic_Signal_Control/Multitime_Period_And_Multischeme_Signal_Control/MyNet.java
package Traffic_Signal_Control.Multitime_Period_And_Multischeme_Signal_Control;

import com.jidatraffic.tessng.*;
import com.jidatraffic.tessng.ColorInterval;
import javax.json.*;
import java.io.*;
import java.util.HashMap;
import java.util.Map;
import java.util.List;
import java.util.ArrayList;

public class MyNet extends JCustomerNet {
    private TessInterface iface;
    private NetInterface netiface;
    private SimuInterface simuiface;
    private Map<Integer, Integer> signalControllerIdMapping = new HashMap<>();
    private Map<Integer, Integer> signalPhaseIdMapping = new HashMap<>();
    private static final String CONFIG_FILE_PATH = "D:\\Yangzhenbei\\Java_demo\\src\\Traffic_Signal_Control\\Multitime_Period_And_Multischeme_Signal_Control\\light_data.json";
    public MyNet() {
        super();
    }

    @Override
    public void afterLoadNet() {
        try {
            iface = TESSNG.tessngIFace();
            netiface = iface.netInterface();
            simuiface = iface.simuInterface();
            // 读取JSON配置文件
            JsonReader jsonReader = Json.createReader(new FileReader(CONFIG_FILE_PATH));
            JsonObject lightData = jsonReader.readObject();
            jsonReader.close();
            System.out.println("读取JSON配置文件成功:" + CONFIG_FILE_PATH);
//            // 创建信号机、信控方案和信号相位
            JsonArray signalControllers = lightData.getJsonArray("signalControllers");
            for (JsonValue controllerVal : signalControllers) {
                JsonObject controllerData = controllerVal.asJsonObject();
                int signalControllerId = controllerData.getInt("id");
                String signalControllerName = controllerData.getString("name");

                // 创建信号机
                ISignalController signalController = netiface.createSignalController(signalControllerName);
                signalControllerIdMapping.put(signalControllerId, (int) signalController.id());
                System.out.printf("创建了信号机，id=%d%n", signalController.id());

                // 处理信控方案
                JsonArray signalPlans = controllerData.getJsonArray("signalPlans");
                for (JsonValue planVal : signalPlans) {
                    JsonObject planData = planVal.asJsonObject();
                    int signalPlanId = planData.getInt("id");
                    String signalPlanName = planData.getString("name");
                    int cycleTime = planData.getInt("cycleTime");
                    int startTime = planData.getInt("startTime");
                    int endTime = planData.getInt("endTime");
                    int offset = planData.getInt("offset");

                    // 创建信控方案
                    ISignalPlan signalPlan = netiface.createSignalPlan(
                            signalController,
                            signalPlanName,
                            cycleTime,
                            offset,
                            startTime,
                            endTime
                    );
                    System.out.printf("创建了信控方案，id=%d%n", signalPlan.id());

                    // 处理信号相位
                    JsonArray signalPhases = planData.getJsonArray("signalPhases");
                    for (JsonValue phaseVal : signalPhases) {
                        JsonObject phaseData = phaseVal.asJsonObject();
                        int signalPhaseId = phaseData.getInt("id");
                        String signalPhaseName = phaseData.getString("name");

                        // 解析信号灯色和时长
                        JsonArray colorsArr = phaseData.getJsonArray("colors");
                        JsonArray durationsArr = phaseData.getJsonArray("durations");
                        ArrayList<ColorInterval> colorIntervals = new ArrayList<>();

                        for (int i = 0; i < colorsArr.size(); i++) {
                            String color = colorsArr.getString(i);
                            int duration = durationsArr.getInt(i);
                            colorIntervals.add(new ColorInterval(color, duration));
                        }

                        // 创建信号相位
                        ISignalPhase signalPhase = netiface.createSignalPlanSignalPhase(
                                signalPlan,
                                signalPhaseName,
                                colorIntervals
                        );
                        signalPhaseIdMapping.put(signalPhaseId, (int) signalPhase.id());
                        System.out.printf("创建了信号相位，id=%d%n", signalPhase.id());
                    }
                }
            }
            System.out.println();

            // 创建信号灯头
            JsonArray signalLamps = lightData.getJsonArray("signalLamps");
            for (JsonValue lampVal : signalLamps) {
                JsonObject lampData = lampVal.asJsonObject();
                String signalLampName = lampData.getString("name");

                // 解析关联的路段ID
                JsonArray linkIdsArr = lampData.getJsonArray("linkIds");
                List<Integer> linkIds = new ArrayList<>();
                for (JsonValue linkVal : linkIdsArr) {
                    linkIds.add(((JsonNumber)linkVal).intValue());
                }

                // 获取信号机ID映射
                int signalControllerId = lampData.getInt("signalControllerId");
                int signalControllerTessId = signalControllerIdMapping.get(signalControllerId);
                ISignalController signalController = netiface.findSignalControllerById(signalControllerTessId);

                // 解析关联的相位ID
                JsonArray phaseIdsArr = lampData.getJsonArray("signalPhaseIds");
                List<Integer> signalPhaseTessIds = new ArrayList<>();
                for (JsonValue phaseVal : phaseIdsArr) {
                    int phaseId = ((JsonNumber) phaseVal).intValue();
                    signalPhaseTessIds.add(signalPhaseIdMapping.get(phaseId));
                }

                // 为每个路段的车道创建信号灯头
                for (int linkId : linkIds) {
                    ILink link = netiface.findLink(linkId);
                    for (ILane lane : link.lanes()) {
                        long laneId = lane.id();
                        double laneLength = TESSNG.p2m(lane.length());
                        double dist = TESSNG.m2p(laneLength - 1);

                        // 创建信号灯头
                        ISignalLamp signalLamp = netiface.createTrafficSignalLamp(
                                signalController,
                                signalLampName,
                                laneId,
                                -1,
                                dist
                        );
                        System.out.printf("创建了信号灯头，id=%d%n", signalLamp.id());

                        // 关联信号相位
                        for (int phaseTessId : signalPhaseTessIds) {
                            netiface.addSignalPhaseToLamp(phaseTessId, signalLamp);
                        }
                    }
                }
            }


        } catch (IOException e) {
            System.err.println("读取配置文件失败: " + e.getMessage());
            e.printStackTrace();
        } catch (JsonException e) {
            System.err.println("解析JSON数据失败: " + e.getMessage());
            e.printStackTrace();
        }
    }



}